lib/cocoapods/installer/user_project_integrator/target_integrator.rb

```diff
@@ -433,6 +433,87 @@ def embed_frameworks_output_paths(framework_paths, xcframeworks)
+          #
+          def add_on_demand_resources(sandbox, project, native_targets, file_accessors, parent_odr_group,
+                                      target_odr_group_name)
+            asset_tags_added = Set.new
+            file_accessors = Array(file_accessors)
+            native_targets = Array(native_targets)
+
+            # Target no longer provides ODR references so remove everything related to this target.
+            if file_accessors.all? { |fa| fa.on_demand_resources.empty? }
+              old_target_odr_group = parent_odr_group[target_odr_group_name]
+              old_odr_file_refs = old_target_odr_group&.recursive_children_groups&.each_with_object({}) do |group, hash|
+                hash[group.name] = group.files
+              end || {}
+              native_targets.each do |user_target|
+                user_target.remove_on_demand_resources(old_odr_file_refs)
+              end
+              old_target_odr_group&.remove_from_project
+              return
+            end
+
+            target_odr_group = parent_odr_group[target_odr_group_name] || parent_odr_group.new_group(target_odr_group_name)
+            current_file_refs = target_odr_group.recursive_children_groups.flat_map(&:files)
+
+            added_file_refs = file_accessors.flat_map do |file_accessor|
+              target_odr_files_refs = Hash[file_accessor.on_demand_resources.map do |tag, resources|
+                tag_group = target_odr_group[tag] || target_odr_group.new_group(tag)
+                asset_tags_added << tag
+                resources_file_refs = resources.map do |resource|
+                  odr_resource_file_ref = Pathname.new(resource).relative_path_from(sandbox.root)
+                  tag_group.find_file_by_path(odr_resource_file_ref.to_s) || tag_group.new_file(odr_resource_file_ref)
+                end
+                [tag, resources_file_refs]
+              end]
+              native_targets.each do |user_target|
+                user_target.add_on_demand_resources(target_odr_files_refs)
+              end
+              target_odr_files_refs.values.flatten
+            end
+
+            # if the target ODR file references were updated, make sure we remove the ones that are no longer present
+            # for the target.
+            remaining_refs = current_file_refs - added_file_refs
+            unless remaining_refs.empty?
+              remaining_refs.each do |ref|
+                native_targets.each do |user_target|
+                  user_target.resources_build_phase.remove_file_reference(ref)
+                end
+                ref.remove_from_project
+              end
+              target_odr_group.recursive_children_groups.each { |g| g.remove_from_project if g.empty? }
+            end
+
+            unless asset_tags_added.empty?
+              attributes = project.root_object.attributes
+              attributes['KnownAssetTags'] = (attributes['KnownAssetTags'] ||= []) | asset_tags_added.to_a
```

Comment:
do we need to remove old asset tags from here?

Reply:
We talked on Slack. The reason I opted out for this is because I wanted to avoid messing the users project in case a tag is shared between the users resources and a pod.